### PR TITLE
Snap 1981

### PR DESCRIPTION
**Pull request by vivekwiz**
## Changes proposed in this pull request
For prepare phase avoid rules that do not handle NullType since that is what is used as placeholder for parameters. Currently for NullType param-literals were getting replaced with null.

## Patch testing
Precheckin

## ReleaseNotes.txt changes
No

## Other PRs 
No

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala [hunk 1]
```diff
@@ -69,21 +70,49 @@ class SnappySessionState(snappySession: SnappySession)
 
   private[sql] var disableStoreOptimizations : Boolean = false
 
+  lazy val analyzerPrepare: Analyzer = new Analyzer(catalog, conf) {
+
+    def getStrategy(strategy: analyzer.Strategy): Strategy = strategy match {
+      case analyzer.FixedPoint(_) => fixedPoint
```
Comment (kneeraj) [inline core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala]: Aren't we losing the maxIterations here. Did not quite understand this.

File: core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala [hunk 2]
```diff
@@ -69,21 +70,49 @@ class SnappySessionState(snappySession: SnappySession)
 
   private[sql] var disableStoreOptimizations : Boolean = false
 
+  lazy val analyzerPrepare: Analyzer = new Analyzer(catalog, conf) {
```
Comment (kneeraj) [inline core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala]: Can you please put a comment on top that what all rules you are skipping and why. Like Promotion rules etc.

File: core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala [hunk 3]
```diff
@@ -69,21 +70,49 @@ class SnappySessionState(snappySession: SnappySession)
...
+      case batch => Batch(batch.name, getStrategy(batch.strategy), batch.rules: _*)
+    }
+
+    override val extendedResolutionRules: Seq[Rule[LogicalPlan]] =
+      getExtendedResolutionRules(this)
+
+    override val extendedCheckRules = getExtendedCheckRules
+  }
+
+  def getExtendedResolutionRules(analyzer: Analyzer): Seq[Rule[LogicalPlan]] =
```
Comment (kneeraj) [inline core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala]: Just to understand. Do we require all these extended rules in the prepare phase? For example does "put into" even works with prepare statement. Do we have a test for some of the scenarios where these rules apply.

File: core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala [hunk 4]
```diff
@@ -69,21 +70,49 @@ class SnappySessionState(snappySession: SnappySession)
...
+        new FindDataSourceTable(snappySession) ::
+        DataSourceAnalysis(conf) ::
+        ResolveRelationsExtended ::
+        AnalyzeMutableOperations(snappySession, analyzer) ::
+        ResolveQueryHints(snappySession) ::
+        (if (conf.runSQLonFile) new ResolveDataSource(snappySession) ::
+            Nil else Nil)
+
+  def getExtendedCheckRules: Seq[LogicalPlan => Unit] =
+    Seq(datasources.PreWriteCheck(conf, catalog), PrePutCheck)
```
Comment (kneeraj) [inline core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala]: Even these PreWriteCheck rules. If my understanding is right these rules will be anyways applied at the execution time. In prepare phase probbaly we can avoid.

File: core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala (hunk 3 above)
Comment (vivekwiz) [reply core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala]: As discussed rules are around for maintainability reason. Updated comment in code.

File: core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala (hunk 4 above)
Comment (vivekwiz) [reply core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala]: Same as above

**Comment by vivekwiz** [comment]
Created SNAP-1987

**Pull request merged**